File: codetraverse/ast_diff/pythondiff.py
import json
from tree_sitter import Language, Parser, Node
import tree_sitter_python
from .Detailedchanges import DetailedChanges
from .basefilediff import BaseFileDiff

class PythonFileDiff(BaseFileDiff):
    """Analyzes and compares two Python ASTs for semantic differences."""
    def __init__(self, module_name=""):
        self.changes = DetailedChanges(module_name)

    def get_decl_name(self, node: Node) -> str:
        """Finds the name of a Python declaration."""
        # For decorated definitions, recurse into the actual definition
        if node.type == 'decorated_definition':
            definition = node.child_by_field_name('definition')
            if definition:
                return self.get_decl_name(definition)

        # For functions and classes, the name is in the 'name' field
        name_node = node.child_by_field_name('name')
        if name_node:
            return name_node.text.decode('utf8')
        
        # For assignments, get the variable name
        if node.type == 'assignment':
            left_node = node.child_by_field_name('left')
            if left_node and left_node.type == 'identifier':
                return left_node.text.decode('utf8')

        return None

    def extract_components(self, root: Node):
        """Extracts all top-level declarations from a Python AST."""
        functions, classes, imports, variables = {}, {}, {}, {}
        
        node_type_map = {
            'function_definition': functions,
            'class_definition': classes,
        }

        for child in root.children:
            node_to_process = child
            
            # Handle decorated definitions by looking at the inner definition
            if child.type == 'decorated_definition':
                definition_node = child.child_by_field_name('definition')
                if definition_node:
                    node_to_process = definition_node
            
            node_type = node_to_process.type

            if node_type in node_type_map:
                name = self.get_decl_name(child) # Use original child to get name from decorator if needed
                if name:
                    target_dict = node_type_map[node_type]
                    target_dict[name] = (child, child.text.decode(errors="ignore"), child.start_point, child.end_point)
            
            elif node_type in ['import_statement', 'import_from_statement']:
                # Record the whole import line as one component
                name = child.text.decode('utf8')
                imports[name] = (child, name, child.start_point, child.end_point)

            # Capture top-level variable assignments
            elif node_type == 'expression_statement' and child.child(0).type == 'assignment':
                assignment_node = child.child(0)
                name = self.get_decl_name(assignment_node)
                if name:
                    variables[name] = (child, child.text.decode(errors="ignore"), child.start_point, child.end_point)

        return functions, classes, imports, variables

    def diff_components(self, before_map: dict, after_map: dict):
        """Compares two dictionaries of components and returns the diff."""
        before_names = set(before_map.keys())
        after_names = set(after_map.keys())
        added_names = after_names - before_names
        deleted_names = before_names - after_names
        common_names = before_names & after_names
        
        added = [(n, after_map[n][1], {"start": after_map[n][2], "end": after_map[n][3]}) for n in sorted(added_names)]
        deleted = [(n, before_map[n][1], {"start": before_map[n][2], "end": before_map[n][3]}) for n in sorted(deleted_names)]
        modified = []
        for name in sorted(common_names):
            _, old_body, old_start, old_end = before_map[name]
            _, new_body, new_start, new_end = after_map[name]
            if old_body.strip() != new_body.strip():
                 modified.append((name, old_body, new_body, {"old_start": old_start, "old_end": old_end, "new_start": new_start, "new_end": new_end}))
        return {"added": added, "deleted": deleted, "modified": modified}

    def compare_two_files(self, old_file_ast: Node, new_file_ast: Node) -> DetailedChanges:
        """The main method to compare two Python files."""
        old_funcs, old_classes, old_imports, old_vars = self.extract_components(old_file_ast.root_node)
        new_funcs, new_classes, new_imports, new_vars = self.extract_components(new_file_ast.root_node)

        category_map = {
            "functions": (old_funcs, new_funcs),
            "classes": (old_classes, new_classes),
            "imports": (old_imports, new_imports),
            "variables": (old_vars, new_vars),
        }
        
        print("/n/n/n/n category map"  ,category_map)

        for category, (old_map, new_map) in category_map.items():
            diff = self.diff_components(old_map, new_map)
            for change_type in ["added", "deleted", "modified"]:
                for item in diff[change_type]:
                    self.changes.add_change(category, change_type, item)

        return self.changes
        
    def process_single_file(self, file_ast: Node, mode="deleted") -> DetailedChanges:
        """Processes a single file that was either entirely added or deleted."""
        funcs, classes, imports, variables = self.extract_components(file_ast.root_node)

        category_map = {
            "functions": funcs,
            "classes": classes,
            "imports": imports,
            "variables": variables,
        }

        for category, component_map in category_map.items():
            for name, data_tuple in component_map.items():
                item = (name, data_tuple[1], {"start": data_tuple[2], "end": data_tuple[3]})
                self.changes.add_change(category, mode, item)
        
        return self.changes

File: codetraverse/ast_diff/pythondiff.ts
import { SyntaxNode } from 'tree-sitter';
import { DetailedChanges } from './Detailedchanges';
import { BaseFileDiff } from './basefilediff';

export class PythonFileDiff extends BaseFileDiff {
    constructor(moduleName: string = "") {
        super(moduleName);
        this.changes = new DetailedChanges(moduleName);
    }

    getDeclName(node: SyntaxNode): string | null {
        if (node.type === 'decorated_definition') {
            const definition = node.childForFieldName('definition');
            if (definition) {
                return this.getDeclName(definition);
            }
        }

        const nameNode = node.childForFieldName('name');
        if (nameNode) {
            return nameNode.text;
        }

        if (node.type === 'assignment') {
            const leftNode = node.childForFieldName('left');
            if (leftNode && leftNode.type === 'identifier') {
                return leftNode.text;
            }
        }

        return null;
    }

    extractComponents(root: SyntaxNode): any[] {
        const functions: any = {};
        const classes: any = {};
        const imports: any = {};
        const variables: any = {};

        const nodeTypeMap: Record<string, any> = {
            'function_definition': functions,
            'class_definition': classes,
        };

        for (const child of root.children) {
            let nodeToProcess = child;

            if (child.type === 'decorated_definition') {
                const definitionNode = child.childForFieldName('definition');
                if (definitionNode) {
                    nodeToProcess = definitionNode;
                }
            }

            const nodeType = nodeToProcess.type;

            if (nodeType in nodeTypeMap) {
                const name = this.getDeclName(child);
                if (name) {
                    const targetDict = nodeTypeMap[nodeType];
                    targetDict[name] = [child, child.text, child.startPosition, child.endPosition];
                }
            } else if (['import_statement', 'import_from_statement'].includes(nodeType)) {
                const name = child.text;
                imports[name] = [child, name, child.startPosition, child.endPosition];
            } else if (nodeType === 'expression_statement' && child.children[0]?.type === 'assignment') {
                const assignmentNode = child.children[0];
                const name = this.getDeclName(assignmentNode);
                if (name) {
                    variables[name] = [child, child.text, child.startPosition, child.endPosition];
                }
            }
        }
        return [functions, classes, imports, variables];
    }

    diffComponents(beforeMap: any, afterMap: any): any {
        const beforeNames = new Set(Object.keys(beforeMap));
        const afterNames = new Set(Object.keys(afterMap));

        const addedNames = Array.from(afterNames).filter(name => !beforeNames.has(name));
        const deletedNames = Array.from(beforeNames).filter(name => !afterNames.has(name));
        const commonNames = Array.from(beforeNames).filter(name => afterNames.has(name));

        const added = addedNames.sort().map(n => [n, afterMap[n][1], { start: afterMap[n][2], end: afterMap[n][3] }]);
        const deleted = deletedNames.sort().map(n => [n, beforeMap[n][1], { start: beforeMap[n][2], end: beforeMap[n][3] }]);

        const modified = [];
        for (const name of commonNames.sort()) {
            const [oldAst, oldBody, oldStart, oldEnd] = beforeMap[name];
            const [newAst, newBody, newStart, newEnd] = afterMap[name];
            if (oldBody.trim() !== newBody.trim()) {
                modified.push([name, oldBody, newBody, { old_start: oldStart, old_end: oldEnd, new_start: newStart, new_end: newEnd }]);
            }
        }

        return { added, deleted, modified };
    }

    compareTwoFiles(oldFileAst: SyntaxNode, newFileAst: SyntaxNode): DetailedChanges {
        const [oldFuncs, oldClasses, oldImports, oldVars] = this.extractComponents(oldFileAst);
        const [newFuncs, newClasses, newImports, newVars] = this.extractComponents(newFileAst);

        const categoryMap = {
            "functions": [oldFuncs, newFuncs],
            "classes": [oldClasses, newClasses],
            "imports": [oldImports, newImports],
            "variables": [oldVars, newVars],
        };

        for (const category in categoryMap) {
            const [oldMap, newMap] = categoryMap[category as keyof typeof categoryMap];
            const diff = this.diffComponents(oldMap, newMap);
            for (const changeType in diff) {
                for (const item of diff[changeType]) {
                    this.changes.add_change(category, changeType, item);
                }
            }
        }

        return this.changes;
    }

    processSingleFile(fileAst: SyntaxNode, mode: string = "deleted"): DetailedChanges {
        const [funcs, classes, imports, variables] = this.extractComponents(fileAst);

        const categoryMap = {
            "functions": funcs,
            "classes": classes,
            "imports": imports,
            "variables": variables,
        };

        for (const category in categoryMap) {
            const componentMap = categoryMap[category as keyof typeof categoryMap];
            for (const name in componentMap) {
                const dataTuple = componentMap[name];
                const item = [name, dataTuple[1], { start: dataTuple[2], end: dataTuple[3] }];
                this.changes.add_change(category, mode, item);
            }
        }

        return this.changes;
    }
}
